### ISSUE TITLE
"Relaxed" mode for qualifiers (e.g. layout, etc...)

### ISSUE THREAD

**Issue by warp-10** [KhronosGroup/glslang#918]
For those of us who implement a "bind by name" policy, it would be nice to omit layout qualifiers completely (e.g. mandatory `input_attachment_index` on `subpassInput` ).

If this feature already exists in some form, feel free to close this issue.

I've tried -auto-map-bindings, but it just causes the help message to be printed.

For the curious: This is because the particular application in question essentially defines all supported renderpass state _before_ loading shaders, and resolves by name (ignoring the qualifiers) instead of failing with a validation error.

**Comment by johnkslang** [comment]
Not sure which layout you are referring to.  "relaxed" implies it is **Offset**, yes?  If **Offset**, Vulkan requires it.  Are you wanting invalid SPIR-V that would then turn into valid SPIR-V?

If **Binding** is the topic, are you getting the usage message just because of no double `-`?  I.e., it needs to be `--auto-map-bindings`, not `-auto-map-bindings` (use `--amb`).

**Comment by warp-10** [comment]
Basically, I want to be able to pass the following to glslangValidator without it generating an error message:

    uniform subpassInput auxColor; // error: requires input_attachment_index
    out vec4 color;
    void main()
    {
        color = subpassLoad(auxColor);
    }

I'm handling the generated SPIR-V directly (option -V). The desired outcome is that it would not generate the corresponding OpDecorate.

The application generates new bytecode with the required decorations, which are determined only from variable names.

**Comment by johnkslang** [comment]
This could possibly be done as a command-line option. My concerns would be this command-line option is a request to:

- not adhere to the Vulkan-GLSL extension
- not generate valid Vulkan SPIR-V
- rely on **OpName** (yes?), even though normally stripping **OpName** is supposed to be a valid transform, but would break your use
- more generally, you could not do any standard SPIR-V transforms (optimizations, compression, etc.) until the SPIR-V is fixed

It would be better to have options that inform the front end on how to generate the valid SPIR-V to begin with.  That's the spirit behind things like `--auto-map-bindings`.

**Comment by warp-10** [comment]
Basically yes on all points.

- The Vulkan driver will receive only valid SPIR-V, though dynamically generated.
- See the above bullet.
- We'd be relying exclusively on `OpName`, and obfuscating symbols for release.
- We'll have our own general-purpose optimizer

Though, It sounds like this might be a more peculiar use-case than I imagined...

Perhaps then just have an option that generates layout/binding indices incrementally, thereby allocating the needed space in the bytecode anyway.

**Reasoning**

The idea is to hide the various complexities introduced by allowing shaders to specify layouts/bindings. When they are ignored by this kind of renderer, they just add visual noise to shader source.

I am also trying to avoid writing a custom frontend by using a subset of standard GLSL.

The rendering engine necessarily knows more about the resources involved and how to schedule their usage than shaders, so it is a maintenance issue to allow artists the ability to mess around with binding points (and a headache to explain why the qualifiers are ignored). We used to allow artists to just write technically "invalid" glsl and then have the rendering engine add the needed decorations to the source text.

As far as optimizations, I was under the impression the driver would be doing the bulk of that work - there really isn't much that can be done with a high-level IL without knowing how it maps to a particular device.
